skbio/metadata/_mixin.py

```diff
@@ -406,3 +407,114 @@ def has_positional_metadata(self):
         """
         return (self._positional_metadata is not None and
                 len(self.positional_metadata.columns) > 0)
+
+
+class IntervalMetadataMixin(metaclass=abc.ABCMeta):
+    def _interval_metadata_axis_len_(self):
+        ''''''
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def __init__(self, interval_metadata=None):
+        raise NotImplementedError
+
+    def _init_(self, interval_metadata=None):
+        if interval_metadata is None:
+            # Could use deleter but this is less overhead and needs to be fast.
+            self._interval_metadata = None
+        else:
+            # Use setter for validation and copy.
+            self.interval_metadata = interval_metadata
+
+    @property
+    @experimental(as_of="0.5.0-dev")
+    def interval_metadata(self):
+        '''``IntervalMetadata`` object containing info about interval features.
+        '''
+        if self._interval_metadata is None:
+            # Not using setter to avoid copy.
+            self._interval_metadata = IntervalMetadata(
+                self._interval_metadata_axis_len_())
+        return self._interval_metadata
+
+    @interval_metadata.setter
+    def interval_metadata(self, interval_metadata):
+        if isinstance(interval_metadata, IntervalMetadata):
+            upper_bound = interval_metadata.upper_bound
+            axis_len = self._interval_metadata_axis_len_()
+            if upper_bound != axis_len:
```

Comment:
I am pretty certain there is an off-by-one error. IntervalMetadata lower/upper bounds are inclusive, so having a lower bound of 0 and an upper bound of 4 (in this example) is five positions, where the number of positions in the Sequence is 4. Here is an example using the `ExampleIntervalMetadataMixin` test object:

```python
In [1]: from skbio.metadata._mixin import IntervalMetadataMixin
In [2]: from skbio.metadata import IntervalMetadata
In [3]: from skbio.util._decorator import overrides
In [4]: # copy and paste the ExampleIntervalMetadataMixin class definition
        # from the unit tests, not included here for brevity

# Create an object of length 4 (i.e. 4 positions). The current code requires
# its IntervalMetadata to have an upper bound of 4, which is inclusive.
In [5]: e = ExampleIntervalMetadataMixin(4, IntervalMetadata(4))

# Confirm the object's axis length is 4 (i.e. 4 positions).
In [6]: e._interval_metadata_axis_len_()
Out[6]: 4

# The IntervalMetadata's lower bound is 0 (inclusive).
In [7]: e.interval_metadata.lower_bound
Out[7]: 0

# The IntervalMetadata's upper bound is 4 (inclusive).
In [8]: e.interval_metadata.upper_bound
Out[8]: 4

# Confirm IntervalMetadata's bounds are inclusive by adding metadata
# that includes the lower and upper bounds (0 and 4, respectively).
In [9]: e.interval_metadata.add(locations=[(0, 4)], metadata={'foo': 42})
Out[9]: Interval(interval_metadata=<4592643040>, locations=[(0, 4)], boundaries=[(True, True)], metadata={'foo': 42})

# Confirm that IntervalMetadata disallows an interval with an upper bound
# of 5.
In [10]: e.interval_metadata.add(locations=[(0, 5)], metadata={'bar': 43})
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-10-2867edfcc7cf> in <module>()
----> 1 e.interval_metadata.add(locations=[(0, 5)], metadata={'bar': 43})

/Users/jairideout/dev/scikit-bio/skbio/metadata/_interval.py in add(self, locations, boundaries, metadata)
    588                         locations=locations,
    589                         boundaries=boundaries,
--> 590                         metadata=metadata)
    591
    592     @_rebuild_tree

/Users/jairideout/dev/scikit-bio/skbio/metadata/_interval.py in __init__(self, interval_metadata, locations, boundaries, metadata)
     80         self._interval_metadata = interval_metadata
     81
---> 82         self._locations_boundaries_setter(locations, boundaries)
     83
     84         # Metadata

/Users/jairideout/dev/scikit-bio/skbio/metadata/_interval.py in _locations_boundaries_setter(self, locations, boundaries)
    228             if boundaries is None:
    229                 locations.sort()
--> 230                 self._check_bounds(locations)
    231                 self._locations = locations
    232                 # reset all the boundaries to True!!

/Users/jairideout/dev/scikit-bio/skbio/metadata/_interval.py in _check_bounds(self, locations)
    252                              'larger than upper bound (%r) or smaller than '
    253                              'lower bound (%r).' %
--> 254                              (locations, upper_bound, lower_bound))
    255
    256     @property

ValueError: Cannot set `locations` ([(0, 5)]) with coordinate larger than upper bound (4) or smaller than lower bound (0).

In [11]:
```